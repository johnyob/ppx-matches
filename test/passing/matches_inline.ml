(* constant patterns *)
let () = assert ([%matches? 42] 42)
let () = assert (not ([%matches? 42] 43))
let () = assert ([%matches? "hello"] "hello")
let () = assert (not ([%matches? "hello"] "world"))

(* variant patterns *)
let () = assert ([%matches? None] None)
let () = assert ([%matches? Some _] (Some 1))
let () = assert (not ([%matches? Some _] None))

(* tuple patterns *)
let () = assert ([%matches? _, 2] (1, 2))
let () = assert (not ([%matches? _, 2] (1, 3)))

(* array patterns *)
let () = assert ([%matches? [| 1; _ |]] [| 1; 2 |])
let () = assert (not ([%matches? [| 1; _ |]] [| 2; 1 |]))

(* record patterns *)
type point =
  { x : int
  ; y : int
  }

let p = { x = 1; y = 2 }
let () = assert ([%matches? { x = 1; y = _ }] p)
let () = assert (not ([%matches? { x = 2; y = _ }] p))

(* or patterns *)
let () = assert ([%matches? 1 | 2 | 3] 2)
let () = assert (not ([%matches? 1 | 2 | 3] 4))

(* when clauses with various conditions *)
let bar = Some 4
let () = assert ([%matches? Some x when x > 2] bar)
let () = assert ([%matches? x when x > 0] 5)
let () = assert (not ([%matches? x when x > 0] (-1)))
let () = assert ([%matches? Some x when x mod 2 = 0] (Some 4))
let () = assert (not ([%matches? Some x when x mod 2 = 0] (Some 3)))

(* range pattern *)
let foo = 'f'
let () = assert ([%matches? 'A' .. 'Z' | 'a' .. 'z'] foo)

(* complex nested patterns *)
let nested = Some (1, [| 2; 3 |])
let () = assert ([%matches? Some (1, [| _; 3 |])] nested)
let () = assert (not ([%matches? Some (2, [| _; 3 |])] nested))

(* lazy patterns *)
let lazy_val = lazy 42
let () = assert ([%matches? lazy _] lazy_val)

(* constraint patterns *)
let () = assert ([%matches? (_ : int)] 42)
