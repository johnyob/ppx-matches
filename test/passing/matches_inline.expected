let () = assert ((match 42 with | 42 -> true | _ -> false)[@warning "-11"])
let () =
  assert (not ((match 43 with | 42 -> true | _ -> false)[@warning "-11"]))
let () =
  assert ((match "hello" with | "hello" -> true | _ -> false)
    [@warning "-11"])
let () =
  assert
    (not ((match "world" with | "hello" -> true | _ -> false)
       [@warning "-11"]))
let () =
  assert ((match None with | None -> true | _ -> false)[@warning "-11"])
let () =
  assert ((match Some 1 with | Some _ -> true | _ -> false)[@warning "-11"])
let () =
  assert
    (not ((match None with | Some _ -> true | _ -> false)[@warning "-11"]))
let () =
  assert ((match (1, 2) with | (_, 2) -> true | _ -> false)[@warning "-11"])
let () =
  assert
    (not ((match (1, 3) with | (_, 2) -> true | _ -> false)[@warning "-11"]))
let () =
  assert ((match [|1;2|] with | [|1;_|] -> true | _ -> false)
    [@warning "-11"])
let () =
  assert
    (not ((match [|2;1|] with | [|1;_|] -> true | _ -> false)
       [@warning "-11"]))
type point = {
  x: int ;
  y: int }
let p = { x = 1; y = 2 }
let () =
  assert ((match p with | { x = 1; y = _ } -> true | _ -> false)
    [@warning "-11"])
let () =
  assert
    (not ((match p with | { x = 2; y = _ } -> true | _ -> false)
       [@warning "-11"]))
let () =
  assert ((match 2 with | 1 | 2 | 3 -> true | _ -> false)[@warning "-11"])
let () =
  assert
    (not ((match 4 with | 1 | 2 | 3 -> true | _ -> false)[@warning "-11"]))
let bar = Some 4
let () =
  assert ((match bar with | Some x when x > 2 -> true | _ -> false)
    [@warning "-11"])
let () =
  assert ((match 5 with | x when x > 0 -> true | _ -> false)[@warning "-11"])
let () =
  assert
    (not ((match (-1) with | x when x > 0 -> true | _ -> false)
       [@warning "-11"]))
let () =
  assert
    ((match Some 4 with | Some x when (x mod 2) = 0 -> true | _ -> false)
    [@warning "-11"])
let () =
  assert
    (not
       ((match Some 3 with | Some x when (x mod 2) = 0 -> true | _ -> false)
       [@warning "-11"]))
let foo = 'f'
let () =
  assert ((match foo with | 'A'..'Z' | 'a'..'z' -> true | _ -> false)
    [@warning "-11"])
let nested = Some (1, [|2;3|])
let () =
  assert ((match nested with | Some (1, [|_;3|]) -> true | _ -> false)
    [@warning "-11"])
let () =
  assert
    (not ((match nested with | Some (2, [|_;3|]) -> true | _ -> false)
       [@warning "-11"]))
let lazy_val = lazy 42
let () =
  assert ((match lazy_val with | (lazy _) -> true | _ -> false)
    [@warning "-11"])
let () =
  assert ((match 42 with | (_ : int) -> true | _ -> false)[@warning "-11"])
